app/controllers/api/v1/project_preferences_controller.rb

```diff
@@ -2,10 +2,34 @@ class Api::V1::ProjectPreferencesController < Api::ApiController
   include PreferencesController
 
   require_authentication :all, scopes: [:project]
+  extra_schema_actions :update_settings
   schema_type :json_schema
+  before_action :validate_records, only: [:update_settings]
+
+  def update_settings
+    upp = UserProjectPreference.find_or_initialize_by(project_id: params_for[:project_id],
+                                                      user_id: params_for[:user_id])
+    upp.settings.merge! params_for[:settings]
+    upp.save
+    render status: :ok, nothing: true
+  end
 
   private
 
+  def find_project
+    Project.find params_for[:project_id] || nil
```

Comment:
I was wrong before, these `.find` methods will raise `ActiveRecord::RecordNotFound` that the API Controller will deal with by default. What you do with this will depend on how you treat the upp find above, if the upp has to exist before you update it then you can proxy the find by user / project id into this lookup, e.g.

``` ruby
UserProjectPreference.find(project_id: params_for[:project_id], user_id: params_for[:user_id])
```
If the upp doesn't have to exist then yeah you should make sure the user / project's can be found, maybe via 
``` ruby
def validate_records
  Project.find(params_for[:project_id])
  User.find(params_for[:user_id]
  unless project.owner?(api_user.user)
    raise Api::Unauthorized, "you must be the project owner"
  end
end
```